### ISSUE TITLE
Support child tables not managed by pg_partman (WAS: run_maintenance fails for legacy partitions with "_p" in the name)

### ISSUE THREAD

**Issue by hopesea** [keithf4/pg_partman#51]
If migrating from a legacy partition management approach to pg_partman, and the existing time-static partitions contain the _p character sequence in their name, run_maintenance fails with ERROR:  invalid value "&lt;substring from legacy partition name&gt;" for "&lt;datetime_string based on partitioning period&gt;" based on partition.

A minimal example to demonstrate the issue:
>CREATE TABLE public.my_parent(id bigserial, ts timestamp without time zone not null, txt varchar(10));
>CREATE TABLE public.my_parent_201511() INHERITS (public.my_parent) WITH (OIDS=FALSE);
>CREATE TABLE public.my_parent_201512() INHERITS (public.my_parent) WITH (OIDS=FALSE);
>CREATE TABLE public.my_parent_201513() INHERITS (public.my_parent) WITH (OIDS=FALSE);
>CREATE TABLE public.my_parent_201514() INHERITS (public.my_parent) WITH (OIDS=FALSE);
>SELECT * FROM partman.create_parent('public.my_parent','ts','time-static',  'weekly', null, 1, true, null, true, false, false); 
>SELECT * FROM partman.run_maintenance('public.my_parent', false, false);

This produces the following error: 

>ERROR:  invalid value "aren" for "IYYY"
>********** Error **********

**Comment by keithf4** [comment]
Yes, you will have to migrate your old partitions to use pg_partman's suffix naming pattern. It's not just any "_p" pattern in the name, it's the final "_p" that it looks for.

For weekly partitioning, the pattern required is: `_pIYYY"w"IW`
Note this is the ISO Year & Week format. The double-quotes around the "w" are required to let postgres know it's not part of the date pattern, but the actual letter "w". So this week would be
```
keith=# select 'my_parent_p'||to_char(now(), 'IYYY"w"IW');
      ?column?      
--------------------
 my_parent_p2015w12
(1 row)
```
The easiest thing to do would be to generate some ALTER statements to rename your tables that decode your old date pattern and rename the table with pg_partman's pattern. Here's an example of one I did for someone else that went from a weekly pattern of `ods_tracking.hits_pYYYY_MMDD` to using the one I gave above. Hopefully you can adapt it to your needs.
```
select 'ALTER TABLE '||n.nspname||'.'||c.relname||' RENAME TO '||substring(c.relname from 1 for 13)||to_char(to_timestamp(substring(c.relname from 14), 'YYYY_MMDD'), 'IYYY')||'w'||to_char(to_timestamp(substring(c.relname from 14), 'YYYY_MMDD'), 'IW')||';'
from pg_inherits h
join pg_class c on h.inhrelid = c.oid
join pg_namespace n on c.relnamespace = n.oid
where n.nspname = 'ods_old'
and h.inhparent::regclass = 'ods_tracking.hits'::regclass
order by c.relname; 
```

I'm working on a legacy migration document to help people with moving their existing partition sets to pg_partman. It will be included in my next release hopefully.

**Comment by hopesea** [comment]
An alternative would be to store or calculate a regex suffix based on the type and interval, and only operate on partitions that match pg_partman's own naming criteria; e.g. a partitioning scheme with type time-static and interval weekly produces datetime_string "IYYY"w"IW". This could be matched with a regex '_p\d{4}w\d{2}$' and only partitions that match that naming convention would be operated against (or returned by function show_partitions). 

Having pg_partman only operate on partitions it has created would eliminate side-effects of legacy partitions and also prevent pg_partman from doing unexpected things (such as the uninheriting of legacy partitions if retention policy is enabled).

**Comment by keithf4** [comment]
That may work ok for the predefined intervals I have, but not so sure about if someone uses time-custom. I could do like you suggest, but I have a feeling I'd be chasing edge cases forever. And having show_partitions() not actually show all the partitions in a set in the proper order would not be the behavior most people expect.

I think the side-effects would be far more unpredictable when mixing existing partition names with ones managed by pg_partman. Long term, it would be best to get things named consistently and renaming them really isn't all that difficult with the query I showed above. I've done it a few times now myself and it's worked fine. If you have any technical concerns other than a preference to keep your old partition names, I'd be glad to hear them out. I can't think of any myself though.

**Comment by hopesea** [comment]
I may be off the mark, but this issue would not only impact legacy partitions, but also the scenario where data growth rates accelerate over time and I need to move from yearly to month, or monthly to weekly partitioning. Would this approach of trying to treat all partitions under a table as uniform and consistent with the current contents of part_config prevent that being an option because the partition name suffix for existing monthly partitions ("YYYY_MM") will not parse into a timestamp once the interval becomes weekly and datetime_string becomes "IYYY"w"IW". Renaming existing partitions is a simple answer by when the interval for those partitions is different to the target interval it's not viable option.

In my case I am moving from a legacy monthly partitioning scheme to a weekly pg_partman partitioning scheme.

**Comment by keithf4** [comment]
Ok. That is a good case. Could also consolidate older data into fewer tables too now that I think about it. I'll see if that pattern matching method works. It will likely be a while before I can get to it, though. I've got a release with a lot of changes in the works right now.

**Comment by hopesea** [comment]
I've submitted a PR demonstrating how I think this could be achieved - creates a simple regex from the datetime_string to constraint which rows are returned by the query produced in show_partitions.

**Comment by keithf4** [comment]
Thanks! I'll try and take a look at it after I get my next release out.

**Comment by keithf4** [comment]
Sorry for the lack of feedback on this. As I've developed more features on pg_partman I've come to realize that supporting child tables in a partition set that are not managed by pg_partman just isn't workable. It's why I have not incorporated your changes into any patches since then because I keep finding more edge cases to handle. I make too many assumptions in my development about what a partition set will look like and trying to work around child tables that don't follow an expected pattern is making more work for me than I'd like. If my work on features ever dies down for this extension, I may come back and look at this. But for now I do not plan on supporting child tables that do not follow pg_partman's default naming patterns.

I've started writing a "migration" howto with instructions on how to convert an existing partition set into one managed by pg_partman. I also do plan on a "consolidation" script that will move data into fewer child tables as time goes on. Hopefully those two things will help with concerns you've brought up here.
